rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,337 @@
+through `push` and `pull` operators.
+
+Making `push` and `pull` top-level commands would streamline UX. For example,
+consider this `vast export` pipeline:
+
+```bash
+# exec
+vast exec 'pull vast://1.2.3.4 |
+           where EXPRESSION |
+           write json'
+# pull
+vast pull 'where EXPRESSION |
+           write json
+```
+
+This assumes that `pull` tries the default VAST endpoint of 1.2.3.4 to connect.
+
+An `import` degenerates to:
+
+```bash
+# exec
+vast exec 'read zeek | where EXPRESSION | push vast:///'
+# push
+vast push 'read zeek | where EXPRESSION'
+```
+
+Note that this hoisting into commands only works because `pull` and `push` are
+head/tail of the pipeline.
+
+### Mutating existing data
+
+Security telemetry data is often messy. After the data collection aspects have
+been sorted out and analysts interact with the data, it becomes clear that some
+part of the data needs to be cleaned. This matches ELT mindset of dumping
+everything first so that the data is there, and then tweak as you go.
+
+Example operations include, migrating data to a different schema, enriching
+telemetry (e.g., with GeoIP data), and fixing invalid entries.
+
+Mutation of data is non-trivial in that it requires exclusive access to the
+data. Otherwise data races may occur. Consequently, we need transactional
+interface (ACID) to support this operation.
+
+As long as VAST owns the underlying data (i.e., only VAST is allowed to make
+changes), VAST can already mutate of data at rest. [Spatial
+compaction][mutate-at-rest] uses a pre-defined disk quota as trigger to apply
+pipelines to a subset of to-be-transformed data. After VAST applies the
+pipeline, VAST optionally removes the original data in an atomic fashion.
+
+[mutate-at-rest]: https://vast.io/docs/use-vast/transform#modify-data-at-rest
+
+Modeled after the `compaction` plugin, we may consider exposing a mutable
+pipeline interface through a dedicate command. This command interacts with a
+node, and could be seen as a combination of `pull` and `push` that executes
+remotely. For example:
+
+```bash
+vast mutate --types=A,B,C 'where 10.0.0.0/8 | put orig.h resp.h'
+```
+
+This would apply the pipeline to filter and project in schema types `A`, `B`,
+and `C`.
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the "interior"
+operators, with only `from` and `to` taking raw input as byte stream. But this
+may be too restrictive. Let's assume that input and output are dynamically
+typed. An operator must then define its input and output types, e.g., `from<In,
+Out>` takes as input `In` and produces instances of `Out`.
+
+### Matcher Example
+
+This can come in handy for working with the `matcher` plugin. Consider the use
+case of constructing a matcher from CSV data, e.g., using this pipeline:
+
+```bash
+vast exec 'from file:///tmp/bad-stuff.csv |
+           read csv |
+           matcher build --extract=net.src.ip' > ips.flatbuffer
```

Comment:
The matcher command is missing from your above table of operators.

Reply:
The `matcher` operator is not really an operator itself. Its sub-commands are. I would rather introduce a dedicated table for matcher subcommands in this section.

Reply:
That doesn't make much sense to me. Isn't the first word after the pipe the operator name?

Reply:
It is. Maybe `matcher` must remain a command then? Otherwise we need to define `load`, `build`, `match`, etc.

Reply:
Nothing's stopping us from having both a `matcher` command and a `matcher` operator. I don't think that's confusing.

Reply:
I fixed this by letting the matcher plugin introduce multiple commands. There's now also a table in the operator synopsis.